NIFI-10381 Refactor Azure Event Hubs components with current SDK

exceptionfactory opened the pull request:
# Summary

[NIFI-10381](https://issues.apache.org/jira/browse/NIFI-10381) Refactors the following Azure Event Hubs components from legacy version 3 to current version 5 of the Azure SDK for Java:

- ConsumeAzureEventHub
- GetAzureEventHub
- PutAzureEventHub

Refactoring also includes moving the `AzureEventHubRecordSink` from the separate `nifi-azure-record-sink-nar` to the shared `nifi-azure-nar`.

With the exception of the `Consumer Host` property in `ConsumeAzureEventHub`, migrating to the current version of the Azure SDK maintains support for current component properties. The current Azure Event Hubs SDK handles unique Event Processor identification using an internal UUID.

Upgrading to the current Azure SDK also provided the opportunity to streamline dependency management using the [Azure SDK Bill of Materials](https://search.maven.org/artifact/com.azure/azure-sdk-bom/1.2.4/pom). The `nifi-azure-processors` module retains legacy dependencies for Azure Storage and Azure Key Vault, but all current Azure SDK libraries can be aligned and managed using the Bill of Materials dependency.

Internal changes include refactor `PutAzureEventHub` to use the standard synchronous client instead of creating a separate executor service. The current Azure SDK maintains an internal elastic executor for handling socket communication.

Additional improvements include standardizing unit tests and removing duplication in some test methods. The `jsr305` module is no longer a transitive dependency of current libraries, so changes include removing this unnecessary exclusion.

# Tracking

Please complete the following tracking steps prior to pull request creation.

### Issue Tracking

- [X] [Apache NiFi Jira](https://issues.apache.org/jira/browse/NIFI) issue created

### Pull Request Tracking

- [X] Pull Request title starts with Apache NiFi Jira issue number, such as `NIFI-00000`
- [X] Pull Request commit message starts with Apache NiFi Jira issue number, as such `NIFI-00000`

### Pull Request Formatting

- [X] Pull Request based on current revision of the `main` branch
- [X] Pull Request refers to a feature branch with one commit containing changes

# Verification

Please indicate the verification steps performed prior to pull request creation.

### Build

- [X] Build completed using `mvn clean install -P contrib-check`
  - [X] JDK 8
  - [X] JDK 11
  - [X] JDK 17

### Licensing

- [ ] New dependencies are compatible with the [Apache License 2.0](https://apache.org/licenses/LICENSE-2.0) according to the [License Policy](https://www.apache.org/legal/resolved.html)
- [ ] New dependencies are documented in applicable `LICENSE` and `NOTICE` files

### Documentation

- [ ] Documentation formatting appears as expected in rendered files

turcsanyip commented:
Reviewing...

turcsanyip reviewed (changes requested):
@exceptionfactory Thanks for upgrading to the new Azure Event Hubs SDK! It is definitely a significant step forward because it is part of the unified Azure SDK and provides a better API. Also, it is a good occasion to get rid of some potential issues in the NiFi code related to synchonization and thread handling/leaking.

I started off with setting up some flows and trying out the updated processors. The `Put` processor works for me. However, I have issues with receiving messages:
- `Get` processor: I think `Message Enqueue Time` is not properly handled. If I set a value, the processor receives the same messages from that point on at every `onTrigger`. If I leave it empty, the processor can skip message published between 2 onTrigger calls. I think the root cause is that `EventPosition` should be determined only once,  in `onScheduled`. The proper behaviour would be to start off from `Message Enqueue Time` or the processor start time (it the property is not specified) and load all messages from that point on while the processor is running.
- `Consume` processor: It seems the processor does not migrate the offsets committed with the previous version of the library but starts with the current offset. Another issue is that if I restart the processor, it will not process anything (only after NiFi restart).

Cloud you please check these issues?

exceptionfactory commented:
Thanks for the feedback and testing @turcsanyip!

I pushed an update incorporating your suggestions for credential handling in `PutAzureEventHub` and also addressed issues with Event Position handling.

For `GetAzureEventHub`, the initial enqueued time determination happens once as suggested, and subsequent executions use the `PartitionProperties` to determine the last enqueued time. This approach should be more robust for handling gaps between processor execution.

For `ConsumeAzureEventHub`, the new  `LegacyBlobStorageEventPositionProvider` is responsible for locating existing checkpoint information and returning it when found. The implementation follows the storage approach used in [AzureStorageCheckpointLeaseManager](https://github.com/Azure/azure-sdk-for-java/blob/main/sdk/eventhubs/microsoft-azure-eventhubs-eph/src/main/java/com/microsoft/azure/eventprocessorhost/AzureStorageCheckpointLeaseManager.java) from the legacy SDK. The implementation will read existing sequence numbers as use them to derive a starting Event Position when found. The provider also deletes the blob since the consumer will subsequently store checkpoint information using the new SDK strategy. This approach should handle the basic migration scenario when upgrading NiFi to a version that includes these changes. It is probably worth including a note in the NiFi migration guidance for awareness.

Please let me know if you notice any additional issues.

turcsanyip commented on nifi-nar-bundles/nifi-azure-bundle/nifi-azure-processors/src/main/java/org/apache/nifi/processors/azure/eventhub/ConsumeAzureEventHub.java:
```diff
@@ -381,313 +366,285 @@ protected Collection<ValidationResult> customValidate(ValidationContext validati
+            eventProcessorClient.start();
 
-        @Override
-        public void onOpen(PartitionContext context) throws Exception {
-            getLogger().info("Consumer group {} opened partition {} of {}",
-                    new Object[]{context.getConsumerGroupName(), context.getPartitionId(), context.getEventHubPath()});
-        }
+            processSessionFactory = sessionFactory;
```
I'd suggest setting `processSessionFactory` before starting the client because the client's event handler may try to use it before the variable gets initialized (almost zero chance though).

turcsanyip commented on nifi-nar-bundles/nifi-azure-bundle/nifi-azure-processors/src/main/java/org/apache/nifi/processors/azure/eventhub/ConsumeAzureEventHub.java:
```diff
@@ -381,313 +366,285 @@ protected Collection<ValidationResult> customValidate(ValidationContext validati
+    @OnStopped
+    public void stopClient() {
+        if (eventProcessorClient != null) {
+            eventProcessorClient.stop();
+            eventProcessorClient = null;
+            processSessionFactory = null;
+            readerFactory = null;
+            writerFactory = null;
```
I recommend calling the `close()` method in a try block and nulling the fields even if the `close() `fails. Otherwise `onTrigger()` can never reinitalize the client because it is not null.

turcsanyip commented on nifi-nar-bundles/nifi-azure-bundle/pom.xml:
```diff
@@ -26,12 +26,9 @@
-        <!-- azure-identity depends on msal4j transitively, keep these versions consistent -->
-        <msal4j.version>1.11.0</msal4j.version>
-        <azure-cosmos.version>4.26.0</azure-cosmos.version>
+        <azure.sdk.bom.version>1.2.4</azure.sdk.bom.version>
```
Could you please bump the bom version to the latest `1.2.6`?
I still have troubles with restarting the Consume processor (no messages processed after restart) but it seems to work with that version.

turcsanyip reviewed (changes requested):
@exceptionfactory I still have issues with the Get and Consume processors. Could you please check my comments?

exceptionfactory commented on nifi-nar-bundles/nifi-azure-bundle/nifi-azure-processors/src/main/java/org/apache/nifi/processors/azure/eventhub/GetAzureEventHub.java:
```diff
@@ -205,98 +193,20 @@ public final List<PropertyDescriptor> getSupportedPropertyDescriptors() {
+    public void onScheduled(final ProcessContext context) {
         final BlockingQueue<String> partitionNames = new LinkedBlockingQueue<>();
         for (int i = 0; i < context.getProperty(NUM_PARTITIONS).asInteger(); i++) {
             partitionNames.add(String.valueOf(i));
```
Thanks for the suggestion, this is a much better user experience. I switched the implementation to use `getPartitionIds()` and set the existing property to not required, with an updated description indicating that it is deprecated.

exceptionfactory commented on nifi-nar-bundles/nifi-azure-bundle/nifi-azure-processors/src/main/java/org/apache/nifi/processors/azure/eventhub/GetAzureEventHub.java:
```diff
@@ -307,95 +217,125 @@ public void onScheduled(final ProcessContext context) throws ProcessException, U
+            final PartitionProperties partitionProperties = eventHubConsumerClient.getPartitionProperties(partitionId);
+            final Instant lastEnqueuedTime = partitionProperties.getLastEnqueuedTime();
+            getLogger().debug("Receiving Events for Partition [{}] from Last Enqueued Time [{}]", partitionId, lastEnqueuedTime);
+            eventPosition = EventPosition.fromEnqueuedTime(lastEnqueuedTime);
```
Thanks for clarifying the issue and pointing out the differences. I updated `GetAzureEventHub` to track last Event Position for each Partition using a `ConcurrentHashMap` that starts with the initial enqueued time in `onScheduled` and subsequently updates the Event Position based on the last Sequence Number observed.

exceptionfactory commented:
Thanks for the additional feedback and recommendations @turcsanyip!

I upgraded the Azure SDK BOM to 1.2.6 and also made changes based on your recommendations. `GetAzureEventHub` now tracks last Event Position per Partition using a `ConcurrentHashMap`. `ConsumeAzureEventHub` also appears to function as designed following multiple restarts when using the new version.

turcsanyip commented:
@exceptionfactory Thanks for the review changes! `Get` works as the previous version now and `Consume` can continue processing after restart.
As the next step, I'm going to test some non-basic use cases like multiple partitions and multiple consumers in a cluster and also to do a thorough review of the code.

turcsanyip reviewed (changes requested):
@exceptionfactory I tested the processors in a clustered environment and found an issue related to the old storage migration (please see my comment below).

I also see random HTTP 409 and 412 error responses on the Consume processor. These are annoying but seem to be harmless (the messages get received in the end). That's why it is not a blocker, just for recording the strange behaviour.
```
com.azure.storage.blob.models.BlobStorageException: Status code 409, "<?xml version="1.0" encoding="utf-8"?><Error><Code>BlobAlreadyExists</Code><Message>The specified blob already exists.
```
```
com.azure.storage.blob.models.BlobStorageException: Status code 412, "<?xml version="1.0" encoding="utf-8"?><Error><Code>ConditionNotMet</Code><Message>The condition specified using HTTP conditional header(s) is not met.
```
It may be related to this issue: https://github.com/Azure/azure-sdk-for-java/issues/29927
However, we use `azure-messaging-eventhub-checkpointstore-blob:1.15.1` where this issue should be fixed.

exceptionfactory commented:
> It may be related to this issue: [Azure/azure-sdk-for-java#29927](https://github.com/Azure/azure-sdk-for-java/issues/29927) However, we use `azure-messaging-eventhub-checkpointstore-blob:1.15.1` where this issue should be fixed.

Thanks for noting the HTTP 409 and 412 messages @turcsanyip, I also noticed that during testing. According to the linked issue, it sounds like that is expected as part of partition load balancing, but it sounds like others have also found it confusing. Something to track for future reference in the Azure SDK and upgrade when a new version is available that changes the behavior.

exceptionfactory commented on nifi-nar-bundles/nifi-azure-bundle/nifi-azure-processors/src/main/java/org/apache/nifi/processors/azure/eventhub/position/LegacyBlobStorageEventPositionProvider.java:
```diff
@@ -0,0 +1,154 @@
+                    final long sequenceNumber = sequenceNumberField.asLong();
+                    eventPosition = EventPosition.fromSequenceNumber(sequenceNumber);
+
+                    blobAsyncClient.delete().block();
```
Thanks for highlighting this issue @turcsanyip.

Part of the challenge with reading the legacy offset information is that the consumer identifiers change between SDK versions. The identifiers are based on a UUID, and the new SDK generates a new UUID.

The other issue is that the Processor needs a way to avoid using the legacy offset information after initial retrieval, as all subsequent handling should be based on the new checkpoint store strategy.  The most straightforward way is to delete the legacy offset information. In this scenario, the first instance of the Processor should read the legacy offset information and proceed. Other instances should begin reading from the latest offset and start tracking using the new checkpoint store strategy. With this approach, at least one consumer will retrieve events based on legacy offset information. It should not be necessary for all consumers to start from previous offsets as long as at least one consumer reads the events from the oldest offset. That might require a small adjustment to this legacy provider.

This is probably something worth noting in migration guidance, but open to additional ideas.

turcsanyip reviewed (approved):
@exceptionfactory Thanks for all the review changes! It looks good to me now.

+1 Merging
